### PR TITLE
ENTERPRISE-696: allow users to specify custom rest-api-path

### PR DESCRIPTION
**Pull request by josf**
Problem
--------

When building the URL, we use base-uri +  "/rest/api/latest". Some
users might have versioned paths, like "/rest/api/2/latest".

Solution
--------

Add a "rest-api-path" config item.

Details
-------

We validate the path by checking for leading (mandatory) and
trailing (illegal) slashes.

We also enforce a protocol (http/https) in base-uri now.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -96,6 +97,7 @@ auth:
     base-uri: https://jira.example.com
     username: jirauser
     password: jirapwd
+	# rest-api-path: /rest/api/latest (default)
```
Comment (jumarko) [inline README.md]: Why is this here in `projects` section?

File: src/codescene_enterprise_pm_jira/jira.clj [hunk 2]
```diff
@@ -5,21 +5,27 @@
...
-                             (-> params
-                                 (assoc-in [:query-params :startAt] start-at)
-                                 (assoc :conn-timeout 5000)))
-          body (:body result)
-          {:keys [issues maxResults]} (json/parse-string body true)]
-      (if (seq issues)
-        (recur (+ maxResults start-at) (concat total-issues issues))
-        total-issues))))
+(defn- resolve-rest-api-path [path]
+  (if (or (nil? path) (string/blank? path))
```
Comment (jumarko) [inline src/codescene_enterprise_pm_jira/jira.clj]: You don't need to check for nil, it's handled by `string/blank?` automatically.

File: src/codescene_enterprise_pm_jira/config.clj [hunk 3]
```diff
@@ -17,9 +17,42 @@
...
 (defn- validate-jira-base-uri
   [jira-base-uri]
-  (when (string/blank? jira-base-uri)
+  (cond
+    (string/blank? jira-base-uri)
     (throw+ {:msg "Missing jira base uri in configuration. Please specify proper value for auth->jira->base-uri key"
-             :type :invalid-config})))
+             :type :invalid-config})
+
+    (not (re-matches #"^https?://.*" jira-base-uri))
```
Comment (jumarko) [inline src/codescene_enterprise_pm_jira/config.clj]: Consider using `java.net.URL` constructor instead for more robust URL validation - validating both 
You can then also remove all those `rest-api-path` validations.

E.g. following URLs will also be considered invalid with `java.net.URL`:
```
  (testing "no host" (check-invalid-jira-uri "https://"))
  (testing "wrong protocol" (check-invalid-jira-uri "https://example.com:XYZ"))
```

Notice, that two (or more) slashes after are usually OK in URLs.

Moreover, I think it'd be better to have just one config key `rest-api-url` which would include everything, but that's backward incompatible :(.

File: README.md [hunk 4]
```diff
@@ -66,6 +66,7 @@ auth:
     base-uri: {uri}
     username: {string}
     password: {string}
+	rest-api-path: {string: optional, defaults to "/rest/api/latest"}
```
Comment (jumarko) [inline README.md]: It seems that this has a wrong indentation.

File: README.md (hunk 1 above)
Comment (josf) [reply README.md]: It should be in the `auth.jira` section... Now I see: tabs instead of spaces. Fixed.

File: README.md (hunk 4 above)
Comment (josf) [reply README.md]: Fixed.

File: src/codescene_enterprise_pm_jira/jira.clj (hunk 2 above)
Comment (josf) [reply src/codescene_enterprise_pm_jira/jira.clj]: Forgot about that. Thanks. Fixed.

**Comment by josf** [comment]
@jumarko I rewrote the URI stuff. Better now?

File: src/codescene_enterprise_pm_jira/config.clj (hunk 3 above)
Comment (josf) [reply src/codescene_enterprise_pm_jira/config.clj]: Interestingly, "https://example.com:XYZ" is a legitimate URI according to java.net.URI. Anyway, it should be much more robust now.

Comment (jumarko) [reply src/codescene_enterprise_pm_jira/config.clj]: I meant `java.net.URL` not `java.net.URI` ;).

**Comment by jumarko** [comment]
Btw. I didn't test anything, just reviewed the code changes.

File: src/codescene_enterprise_pm_jira/config.clj (hunk 3 above)
Comment (josf) [reply src/codescene_enterprise_pm_jira/config.clj]: Ha! Ok, now I've fixed that too, so we should have really robust uri handling now.
